### PR TITLE
feat: Add authentik as a OAuth provider

### PR DESCRIPTION
**Pull request by Aragur**
## What does this PR do?

This PR adds [authentik](https://goauthentik.io/) as a authentification provider 🚀.   
This is a killer feature for me, as this is the only *self-hosted* OAuth provider 🔥 .

## Test Plan

I tested signing in with this [OAuth-Tester](https://github.com/AmreshSinha/OAuth-Tester-Appwrite).  
Everything worked  as intended.

## Related PRs and Issues

None 😄 (If required I may create an Feature-Request for this)

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

Yes I did.

### CODE REVIEW

**Comment by Aragur** [comment]
@eldadfux Is there something wrong with my PR or why does nobody review or comment it?

**Comment by eldadfux** [comment]
The team is pretty busy, we'll review it ASAP before the release of the next Appwrite version.

File: src/Appwrite/Auth/OAuth2/Authentik.php [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+     /**
+     * Extracts the authentik Domain from the JSON stored in appSecret
+     * 
+     * @return string
+     */
+    protected function getAuthentikDomain(): string
+    {
+        $secret = $this->getAppSecret();
+        return (isset($secret['authentikDomain'])) ? $secret['authentikDomain'] : '';
```
Comment (christyjacob4) [inline src/Appwrite/Auth/OAuth2/Authentik.php:211]:
```suggestion
        return $secret['authentikDomain'] ?? ''; 
```

File: src/Appwrite/Auth/OAuth2/Authentik.php [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+        if (isset($user['email'])) {
+            return $user['email'];
+        }
+        
+        return '';
+    }
+
+    /**
+     * Check if the User email is verified
+     *
```
Comment (christyjacob4) [inline src/Appwrite/Auth/OAuth2/Authentik.php:142]:
```suggestion
```

File: src/Appwrite/Auth/OAuth2/Authentik.php [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+    /**
+     * Extracts the Client Secret from the JSON stored in appSecret
+     * 
+     * @return string
+     */
+    protected function getClientSecret(): string
+    {
+        $secret = $this->getAppSecret();
+
+        return (isset($secret['clientSecret'])) ? $secret['clientSecret'] : '';
```
Comment (christyjacob4) [inline src/Appwrite/Auth/OAuth2/Authentik.php:200]:
```suggestion
        return $secret['clientSecret'] ?? ''; 
```

**Review by christyjacob4 — CHANGES_REQUESTED** [review]
Looks awesome! Just some minor comments :)

**Comment by christyjacob4** [comment]
@Aragur please sync with master, the tests are failing due to a missing CNAME.

**Comment by christyjacob4** [comment]
@Aragur So close to merging 😉 
Please check for the code formatting issue raised by the Linter and then we're good to merge. 
Can you also share some screenshots of the Users->Settings screen with the Authentik provider added and a screenshot of a session created with the authentik provider ?

**Comment by Aragur** [comment]
> @Aragur So close to merging 😉 Please check for the code formatting issue raised by the Linter and then we're good to merge. Can you also share some screenshots of the Users->Settings screen with the Authentik provider added and a screenshot of a session created with the authentik provider ?

Thanks for you reply.  
The linting task wasn't there once I originaly created my pull request :) But it should be fixed now.

Here are the requested screenshots:
Users->Settings:
Screenshot 2022-06-13 at 21 24 52
Users->Settings->authentik-Details-Screen
Screenshot 2022-06-13 at 21 25 14
OAuth-Screen:
Screenshot 2022-06-13 at 21 29 28
OAuth-Session-Details by [OAuth-Tester-Appwrite](https://github.com/AmreshSinha/OAuth-Tester-Appwrite)
Screenshot 2022-06-13 at 21 29 36

**Comment by Aragur** [comment]
So I modified all required steps 13 days ago...  
I DON'T want to resolve those merge conflicts again which only exists because of the lack of your communication.  
How should an OpenSource project work if you take over 2 months to accept a PR...

**Comment by MoritzHeld** [comment]
I would really appreciate if you guys could accept this PR. Would love to use that feature in a project of mine!

**Comment by Aragur** [comment]
@christyjacob4 Merge is done. All checks ran successfully. What to do next?

File: app/config/providers.php [hunk 4]
```diff
@@ -31,6 +31,16 @@
         'beta' => false,
         'mock' => false,
     ],
+    'authentik' => [
+        'name' => 'authentik',
```
[the remark this answers is not in the document]
Comment (Aragur) [reply app/config/providers.php:35]: So this is the default name (written with a small a).
You can see it on the main page of the docs: https://goauthentik.io

But if it's required by your guidelines, I of course can change it.

**Comment by Aragur** [comment]
> One last comment

I don't know if you saw my comment on your comment. So here it is again:

So this is the default name (written with a small a).
You can see it on the main page of the docs: https://goauthentik.io/

But if it's required by your guidelines, I of course can change it.

**Comment by christyjacob4** [comment]
@Aragur  Sorry for the delay. Yes, our guidelines require the OAuth Provider names to be in title case so as to look consistent with the other providers in the UI

**Comment by Aragur** [comment]
> @Aragur Sorry for the delay. Yes, our guidelines require the OAuth Provider names to be in title case so as to look consistent with the other providers in the UI

Okay, It's done now

**Review by christyjacob4 — APPROVED** [review]
Looks amazing! 🥳  Thank you for your help and patience!

**Comment by lost-RD** [comment]
@Aragur please can you share your version of the OAuth-Tester-Appwrite repo? I'm struggling to get the auth flow right.

**Comment by Aragur** [comment]
@lost-RD here you go: https://github.com/AmreshSinha/OAuth-Tester-Appwrite

**Pull request merged**